### PR TITLE
Selective unfolding

### PR DESCRIPTION
**Pull request by Lionel-Rieg**
This PR introduces the possibility to unset the printing of notations in a more fine grained fashion than Unset Printing Notations. One can unfold notations either individually or by scopes.
It is a project I started at the coding sprint last year and I finished it in the train coming back from this year's.
This is very useful for debugging purposes when having very large notations that should not be unfolded (UTF8 string comes to mind immediately).

It introduces four new commands:
- Deactivate Notation Scope scope_name.
- Reactivate Notation Scope scope_name.
- Deactivation Notation notation_string [From Scope scope_name].
- Reactivation Notation notation string [From Scope scope_name].

Notation strings are the displayed notations with variables replaced by "_".

I agree the names are not so great and could be improved (Hugo suggested "Unset Printing Notation Scope <scope_name>" if i remember correctly).

### CODE REVIEW

**Comment by JasonGross** [comment]
I am strongly in favor of this commit.  I agree with Hugo, the syntax should be:
```coq
(Unset|Set) Printing Notation Scope[s] scope_name ... scope_name.
(Unset|Set) Printing Notation notation_string [From Scope scope_name].
```

File: doc/refman/RefMan-syn.tex [hunk 1]
```diff
@@ -321,6 +321,20 @@ \subsection{Displaying symbolic notations}
 To do so, the option {\em only parsing} is allowed in the list of modifiers of
 \texttt{Notation}.
 
+\paragraph{Deactivating some printing notations}
+In addition \verb=Unset Printing Notations=, it is possible to deactivate notations in a more fine-grain way, either individually or by scopes, using the following syntax:
+\begin{quote}
+  Deactivate Notation "expression" From Scope {\scope}.
+  Reactivate Notation "expression" From Scope {\scope}.
```
Comment (JasonGross) [inline doc/refman/RefMan-syn.tex]: If you permit leaving off the scope qualifier, you should mention it here.

**Comment by Lionel-Rieg** [comment]
The scope qualifier can only be omitted when there is no scope attached to the notation, ie the so-called "lonely notations".  I will fix the refman once we agree on the correct syntax to use (and I also need to be more precise on the fact that it only deactivates printing, not parsing).

Another topic of discussion would be whether invalid uses should be warnings or errors.
Currently, deactivating a notation that does not exist triggers an error whereas all the other misuses (deactivating an already inactive notation and similarly for reactivation) only produce a warning.

Would a command showing inactive notations be useful? I already have code for this (for debugging purposes) but I did not make it available to the user.

**Comment by gares** [comment]
I think @herbelin is the most knowledgeable to review this PR

**Comment by herbelin** [comment]
Let me give a couple of comments:

On the feature side:
- Jason's proposal for a syntax looks to me good, even if it would not enter the standard "Set" mechanism and hence still require an ad hoc vernac entry.
- There is the usual question: do we want these commands to be synchronous with backtrack or not? It depends on whether we intend them to be part of a script (hence synchronized) or part of a communication between an interface and coq (hence stored at the interface level only). I don't remember the conclusion of Pierre Courtieu on this issue. Wasn't it that we need to get two commands, like "Set blabla" for the synchronized one and "Stm Set blabla" for the interface-side one? In any case, it seems to me that the commands in this PR should be synchronized and that declare_summary and declare_object should be used (or, alternatively, the Goptions.MakeStringTable interface which does that automatically, but which might be a bit rigid to use - see examples in detyping.ml).

On the syntactic side:
- This is relatively minor, but I would not charge too much constrextern.ml (from interp, and primarily devoted to the glob_constr / constr_expr translation) with code managing tables; I would move the  functions interpreting the commands to metasyntax.ml, where other toplevel commands are already managed, and the functions managing the tables in notation.ml where other kinds of able are already managed.
- It looks preferable to me that all names in a given file have a common form (i.e. all printing options having a name of the form either print_foo or printing_foo but preferably not a mix of them).
- It looks to me good practice not to change indentation in a part of a file which has no semantic changes: it makes studying commits easier, allowing to concentrate on the real semantic changes. If a syntactic change (alpha-conversion, indentation, improved layout, comments, ...) is justified, it should better go in a separate commit.
- If possible, rework the local history of commits in semantic units. Here, I guess that all can go in a single commit, without having to tell what were the different draft or debugging development steps.
- To whom will do the merge, I don't know if I'm the only one in feeling this but we would get a more directly informative message if the commit message where not just "merge pr/XXX" but if it would also remind in a few words what "XXX" is referring to.

Lionel, if ever you want we meet to discuss more into details of this (useful!) contribution, e.g. for the declare_summary/declare_object stuff, don't hesitate.

And a question to Enrico in passing: what is the current status of texmacspp.ml and of DoCoq?

**Comment by Lionel-Rieg** [comment]
Thanks Hugo for the feedback.
As this is my first try at a pull request, there is definitely room for improvement, notably on the syntactic side, as you noted.  If we have time to discuss this in Montevideo, it would be great.
On the feature side, to me this PR is mostly a debugging tool, so probably not meant to be synchronized.  To me, the normal way to inactivate notations should be by scopes (although they can still be used for printing with delimiter keys).

Maybe a more interesting (and ambitious) change would be to have globally a finer control over notations, allowing for instance to separately define them, activate them (either for parsing or printing), add/remove them to/from scopes, etc. But this is way too much for me to tackle I am afraid.

**Comment by maximedenes** [comment]
@herbelin what should be done with this PR?

**Comment by herbelin** [comment]
@maximedenes: It seems to be a feature which some people wants (and not only for debugging), either in this form or in the more fine-grained form later suggested by @Lionel-Rieg. So, to me,the question is somehow whether Lionel has time to work on it (or to work on it with me - or maybe with @ppedrot, either remotely, or when he will be next time in Paris).

**Comment by Lionel-Rieg** [comment]
I must admit I have not taken another look at this in a long time.
I probably will not have time in the next weeks to start working on this again so feel free to do it without me if you want.  Improving the current version should be pretty straightforward and fast but I still think that a finer control over notations would be useful.  If we go this route, then there are also some design questions that should be answered.  Maybe it is worth a CEP?

**Comment by Lionel-Rieg** [comment]
I have a new version of this PR that integrates most of the comments above.  As I reimplemented it from scratch, it is probably simpler to create a new PR and close this one.  Any strong opinion against that?

**Comment by maximedenes** [comment]
> I have a new version of this PR that integrates most of the comments above. As I reimplemented it from scratch, it is probably simpler to create a new PR and close this one. Any strong opinion against that?

Sorry @Lionel-Rieg, somehow your message went below my radar (and not just mine, apparently). Reopening a PR is fine, force pushing this one is also fine. The feature freeze for 8.7 is June 1st, but if you submit before that, we will try to integrate it. Thanks!

**Comment by Lionel-Rieg** [comment]
I just created PR #703 to replace this one.

**Pull request closed without merging**